Implementation of Issue #4044. Added ScientificTable and ScientificCell subclasses.

dkua opened the pull request:
This implements the requested feature in issue #4044 

Two subclasses, `ScientificTable` and `ScientificCell`, were created of `Table` and `Cell` respectively. An extra kwarg was added to the `table()` factory function for choosing the type of `Table` to use.

The factory function and the table types were tested with the `test_table_types` test in `test_table.py`. A new test image was also added. The test passes in Ubuntu 12.04 64-bit.

The feature was implemented by having `ScientificCell` extend `Cell` with an additional method `get_path()` which simply returns a new `Path` instance that doesn't draw a line when moving between the vertices that make up a vertical line.

This works because `Cell.draw()` calls `Rectangle.draw()` which calls `self.get_path()`.  However `Cell` passes into `Rectangle.draw()` the arguments `self, renderer`. So the `self.get_path()` call is actually calling `Cell` but since it doesn't have that method it uses it's super class' definition which is `Rectangle`. By creating the new subclass with our own `get_path()` method we can implement this feature while following the same logic.

tacaswell commented:
We now have two implementations of this https://github.com/matplotlib/matplotlib/pull/4245

tacaswell commented:
Looking at this PR and #4245, I like the implementation of the new `Cell` type is this PR and the cell type selection mechanic from other PR.

dkua commented:
Thanks for responding so quickly @tacaswell. The newest commit removes the Table subclass and implements a similar cell type selection mechanic as you liked.

dkua commented:
That's strange. Travis failed this single test: https://travis-ci.org/matplotlib/matplotlib/jobs/55347259#L6121

But running the `test_lines` and `test_table` myself on my system, they pass.

tacaswell commented:
That failure is just travis being flaky.  That particular test is verifying that an optimization actually makes things better.  If the VM chokes (due to a greedy neighbour or other VM related things) during the timing of the optimized section the test will fail.  

I kicked it to restart the 2.7 tests.

dkua commented:
Tests pass! Thanks @tacaswell

stallone1993 commented:
@tacaswell We've spoken to the groups who contributed the other PRs. We've agreed that this PR provides a sound implementation that we are all happy with. The other PRs have been closed while this remains open - any updates on potentially merging this feature? Thanks!

tacaswell commented:
I should have time this weekend unless someone beats me to it.

stallone1993 commented:
@tacaswell - thanks for the heads up. Cheers!

efiring commented:
Just a quick thought: I suggest you reconsider the names.  This is really about two alternative styles, neither of which has anything to do with science.  Fundamentally, a table can have horizontal and/or vertical dividers, or no dividers, or dividers in some locations and not in others.  If we are never going to have options other than whether to use vertical dividers, then the styles might be "open cells" (what you are calling "Scientific" versus "closed cells".  But it might be good to look farther ahead, and lay the groundwork for a more complete set of formatting options.

dkua commented:
@efiring Maybe the term `VerticalCell` and `HorizontalCell` would be more apt terms respectively for cells with only vertical and only horizontal lines?

efiring commented:
Yes, with a cell-based approach, that would be better.  But might it make sense to simply have one Cell class that can put lines on any or all of the sides?  The general specification would be a string, e.g. 'LRBT' for all sides, 'BT' for bottom and top, etc.  Aliases could be 'open' for no lines (''), 'closed' for all sides, 'horizontal' for horizontal lines ('BT'), and 'vertical' for vertical lines ('LR').

dkua commented:
@efiring Wouldn't the Cell-based approach potentially be simpler in the case that the logic for each type of Cell is more complex than simply swapping the Path instance? Instead of having all of that logic inside of a single Cell class we have multiple subclasses of Cell. Do end-users interact with the Cell types directly?

efiring commented:
No, I expect users will interact with the table, not the Cell types.  But I also expect that the next request from users will be for the ability to control the cell type on a cell-by-cell basis, not just for the table as a whole.  Based on a *very* quick look, I think one could allow (but not require) an array of cell types, just as there is an array of text entries.  Either way, the Table is drawing cells with properties.  If the Cell (or a single FancyCell subclass) accepts a boxstyle property (or perhaps even a dictionary of properties), then there is no need for more and more subclasses.
I hope I'm not leading you astray; I have never used the table or studied the code.  But I'm concerned about locking in a new API (public classes, etc.).  As mpl matures, it gets harder and harder to modify the API, so more up-front design thought is warranted.  I recognize that this makes contributions like this more difficult.

dkua commented:
@efiring At the moment there isn't a way for users to declare exactly what kind of Cell they want at (x, y) position. It seems that for a single Table, one type of Cell is used throughout. It might be a premature optimization to give the Table class and table() factory function the ability to declare what Cell is used for what position. Maybe it's best left for some future point in time to change the API for selecting separate kinds of Cells.

efiring commented:
Agreed; I'm not arguing that you need to do all of that, but rather that what you do now should leave the door open as much as possible.  That's why I am thinking in terms of having a minimum number of Cell subclasses (like one) instead of adding on the specialized HorizontalCell and VerticalCell (or just one of these).  Once they exist, they have to be supported for quite a while.

zairmubashar commented:
@efiring We've tried to change the code to work as you've suggested. [Here](https://github.com/matplotlib/matplotlib/compare/master...dkua:feature_4044_fancy#diff-cd21097aa8cce79ca8a23b7f52c48f99R233) is the new code.

To provide a quick summary:
- An arguement to the `table()` function to specifies which cell lines should show.
(for example, "LRBT" (in any order) creates a table where each cell has left, right, bottom, and top boundary visible. All aliases that were suggested are also implemented.)
- The `Table` instance stores the above information.
- With creation of a new cell instance from `add_cell()` in `Table` (the cell is now a `FancyCell` - A subclass of `Cell`), the cell is passed in the mentioned "LRBT" type information so that the path returned by `get_path()` allows for the appropriate lines to be in/visible. In the future, if individual cell edges need to be made in/visible it may be done so in `FancyCell`. Of course, this is because the cell instance stores its own "LRBT" type information.

Please let us know if this is what you had in mind.

efiring commented:
Yes, that is what I had in mind, thank you.  It looks like you might be able to factor out the validation of the edge letter string; typically, if there is a setter, then this can be used in the __init__ to handle validation.  Also, for the "closed" or "LRBT" case, maybe you need to just use the rectangle; one way or another, the polygon should be closed so that when it is drawn there is no line-end artifact at the starting/ending corner.

tacaswell commented:
I also like the new version much better (but have not read it in detail).

dkua commented:
@tacaswell @efiring Cleaned up some of the code created by @zairmubashar and added it to this PR. Please have a look and let us know what you think.

efiring commented on lib/matplotlib/table.py:
```diff
@@ -146,6 +147,57 @@ def set_text_props(self, **kwargs):
+
+    """
+
+    EDGES = 'BRTL'
```
I don't think this is needed; it is used in only one place, two lines below here.

dkua replied:
The EDGES variable is also called in the __init__() and get_path() methods.

efiring replied:
Right, I was confused, and had to amend my comment.

efiring commented on lib/matplotlib/table.py:
```diff
@@ -146,6 +147,57 @@ def set_text_props(self, **kwargs):
+    """
+
+    EDGES = 'BRTL'
+    EDGE_ALIASES = {'open':         '',
```
Variable name should be lower case.

dkua replied:
Should they be prepended with an underscore like "_edges"? To signify that they shouldn't be accessed or should they be accessible?

efiring replied:
I think that a leading underscore would be fine.  They really are implementation details.

efiring commented on lib/matplotlib/table.py:
```diff
@@ -146,6 +147,57 @@ def set_text_props(self, **kwargs):
+        else:
+            for edge in visible_edges:
+                if edge not in self.EDGES:
+                    msg = 'Invalid edge param {0}, must only be one of {1} or \
```
We try hard to avoid trailing backslashes.  Maybe this can be avoided here by using parentheses and implicit string concatenation.

dkua replied:
`("something {0}", "something {1}").format(...)` doesn't work as it's a tuple.

`ValueError("something", ..., "something", ...)` seems to just raise it as `ValueError: ("something", ..., "something", ...)` and doesn't implicitly concatenate the strings.

efiring replied:
Try
```python
('abc{0}'
'def{1}').format(1, 2)
```

dkua replied:
That works, didn't know about that. Will switch to using that.

efiring commented on lib/matplotlib/table.py:
```diff
@@ -146,6 +147,57 @@ def set_text_props(self, **kwargs):
+        'Return a path where the edges specificed by _visible_edges are drawn'
+
+        codes = [Path.MOVETO] * 4
+        codes.append(Path.CLOSEPOLY)
```
It looks like this gets overwritten by LINETO even when all edges are visible.

dkua replied:
Do you mean it gets overwritten to `LINETO` when all edges are invisible? We want it to be overwritten in the case all edges are visible.

The `MOVETO`s are only overwritten by the `LINETO` when specified in `self._visible_edges`. (in lines 191 and 192)

efiring replied:
When all edges are visible, I think that CLOSEPOLY is what makes for a seamless boundary, with each corner rendered the same way.  If you overwrite it with LINETO, I think you will have a corner consisting of the start of the boundary and the end of the boundary, but not properly joined.  What path do you get from Rectangle?  That is what you should get from the 'closed' case here.

dkua replied:
Hmm I see what you mean. We've actually hit a small bug with `CLOSEPOLY` in some extra testing we're doing and will report back with news soon/

efiring commented on lib/matplotlib/table.py:
```diff
@@ -216,13 +269,21 @@ def add_cell(self, row, col, *args, **kwargs):
         self._cells[(row, col)] = cell
 
+    @property
+    def drawLines(self):
```
Throughout, I think `drawLines` would work better as `edges`.

dkua replied:
So you'd like the edge logic to be in `drawLines`?

efiring replied:
Sorry, I wasn't clear.  I was referring only to the name. `drawLines` sounds like a command--so it evokes the name of a function that will draw something.  But what it really is is just a specification of edges to be drawn, so just plain `edges` seems like a clearer label.

dkua replied:
Okay gotcha

dkua commented:
@efiring we've implemented your suggestions in the latest commit. Travis failed weirdly on the Python 3.4 VM, can we get a restart on that?

efiring commented on lib/matplotlib/tests/test_table.py:
```diff
@@ -79,3 +79,25 @@ def test_label_colours():
               loc='best')
+
+
+@image_comparison(baseline_images=['table_cell_manipulation'],
```
Add the `remove_text=True` kwarg here to simplify the figures, and reduce the likelihood of spurious differences related to font handling.

tacaswell replied:
This has been addressed.

efiring commented on lib/matplotlib/tests/test_table.py:
```diff
@@ -79,3 +79,25 @@ def test_label_colours():
+    fig = plt.figure(figsize=(4, len(cells)+1))
+    for i, cell in enumerate(cells, 1):
+        ax = fig.add_subplot(8, 1, i)
+        ax.plot([1, 2, 3, 4, 5, 6])
```
What is the rationale for including this blue line?  If it is not essential, better to leave it out.

dkua replied:
The test case was made from the code used to reproduce the issue.

efiring commented on lib/matplotlib/tests/test_table.py:
```diff
@@ -79,3 +79,25 @@ def test_label_colours():
+                rowLoc='center',
+                edges=cell,
+                )
+    plt.tight_layout()
```
This isn't essential to the test, is it?

efiring commented on lib/matplotlib/tests/test_table.py:
```diff
@@ -79,3 +79,25 @@ def test_label_colours():
+    l2 = ('DLI', 35, 38, 10, 22, 25, 85, 22)
+    cells = ('horizontal', 'vertical', 'open', 'closed', 'T', 'R', 'B', 'L')
+
+    fig = plt.figure(figsize=(4, len(cells)+1))
```
The typical way of doing this now is
```python
fig, axs = plt.subplots(nrows=len(cells), figsize=(4, len(cells) + 1))
for ax in axs:
    ax.table(...)
```

efiring commented on lib/matplotlib/table.py:
```diff
@@ -455,23 +526,24 @@ def get_celld(self):
           rowLabels=None, rowColours=None, rowLoc='left',
           colLabels=None, colColours=None, colLoc='center',
-          loc='bottom', bbox=None)
+          loc='bottom', bbox=None, edges='BRTL')
```
Here and above, maybe it would be better to use 'closed' instead of the perhaps initially puzzling 'BRTL'.

efiring commented:
This is getting close.

dkua commented:
@efiring I've added changes to the test that you wanted and changed the default edges type to "closed".

efiring commented:
Thank you.  @tacaswell or @weathergod, any more suggestions?
I think this is ready for entries in doc/users/whats_new/ and doc/api/api_changes/.  In both cases there are a README and the existing entries to use as examples. The reason for using individual files for these entries is that it greatly reduces merge conflicts.

zairmubashar commented:
@tacaswell @WeatherGod, are there any more suggestions or should we pull together some entries for the docs?

OceanWolf commented:
I haven't had a good look though this yet, but from what I see, it looks like we should use this to implement the ``Legend`` class.  At the moment Legend does all the calculations by hand, something I think the ``Table`` class should take care of.

Does ``Legend`` currently require any extra features not currently addressed by ``Table``?

dkua commented:
@OceanWolf we're not sure. We can look into it. But might that not be a topic for another issue/PR?

OceanWolf commented:
@dkua sure thing, I only ask now to double check on the ``Table`` API.

tacaswell commented:
@OceanWolf Overhauling `Legend` is definitely out of scope for this PR!

tacaswell commented on lib/matplotlib/table.py:
```diff
@@ -146,6 +147,67 @@ def set_text_props(self, **kwargs):
+    """
+
+    _edges = 'BRTL'
+    _edge_aliases = {'open':         '',
```
trim the extra white space please.

We much not have pep8 compliance on the rest of this file or the test would be failing.

dkua replied:
@tacaswell do you mean the extra whitespace inside the _edge_aliases dictionary? I can't see any other whitespace and yeah pep8 isn't failing.

tacaswell replied:
I would expect pep8 to require

```python
_edge_aliases = {'open': '',

```

dkua replied:
@tacaswell afaik spaces are allowed after colons inside of dictionaries.

Nothing against it is mentioned in the PEP 8 style guide: https://www.python.org/dev/peps/pep-0008/

And running pep8 on a minimal example and people on SO seem to confirm: http://stackoverflow.com/questions/13497793/python-alignment-of-assignments-style

I'm also at PyCon right now, I could ask people here too

tacaswell replied:
Well, learn something every day.

tacaswell commented:
:+1:  Looks good!  I like the path-based solution to drawing the edges.

@efiring I think this just needs a whats_new, not an api_changes as I don't think it breaks any existing API.

efiring commented:
@tacaswell, good point; I was confused as to when an api_changes entry is needed.

dkua commented:
@tacaswell @efiring could you point us to where (or how) we would write up a whats_new entry?

tacaswell commented:
https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new

The readme in that folder as an explanation.

On Sun, Apr 12, 2015 at 11:38 AM David Kua <notifications@github.com> wrote:

> @tacaswell <https://github.com/tacaswell> @efiring
> <https://github.com/efiring> could you point us to where (or how) we
> would write up a whats_new entry?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4259#issuecomment-92082159>
> .
>

tacaswell commented:
I think this is waiting on a what's new entry and that PEP8 issues?

dkua commented:
@tacaswell yes it seems just those two issues

tacaswell commented:
I am apparently confused about PEP8, so just one issue left :smile:

WeatherGod commented:
@tacaswell, I could have sworn the same thing until someone called me out
on it on a code review last month. I went to look it up and I couldn't find
the entry that I could have sworn was there explaining the formatting of a
dictionary. Maybe it changed at some point?

On Mon, Apr 13, 2015 at 9:45 PM, Thomas A Caswell <notifications@github.com>
wrote:

> I am apparently confused about PEP8, so just one issue left [image:
> :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4259#issuecomment-92555926>
> .
>

dkua commented:
@tacaswell I've added a What's New entry to the docs. We've also added an extra tests for the Cells themselves just for extra assurance that things work.

tacaswell commented:
@dkua @zairmubashar Thanks!

dkua commented:
@tacaswell @efiring thank you very much.
